File: src/bridges/mock.ts
/**
 * Mock bridge: deterministic, schema-driven outputs with latency and failure injection.
 * Used for tests, `gren run --bridge mock`, and to prove failure domains / retries / quorum behaviour
 * without spending a token.
 *
 * Schema hints:
 *   "examples": [..]        pick one (seeded)
 *   "x-mock": <value>        use exactly this value
 *   "enum": [..]             pick one (seeded)
 * Verify role: verdict is "kill" with probability `killRate` (default 0.25).
 */
import crypto from "node:crypto";
import { estimateCostUsd } from "../models.js";
import { BridgeError, type AgentRequest, type AgentResponse, type Bridge, type BridgeContext } from "./types.js";

export interface MockOptions {
  latencyMs?: number;
  jitterMs?: number;
  /** Probability [0,1] that an attempt throws a transient error. */
  failRate?: number;
  /** Probability that an attempt returns schema-invalid output (exercises the repair loop). */
  invalidRate?: number;
  killRate?: number;
  seed?: string;
  /** Node ids that always fail (to demonstrate failure domains). */
  alwaysFail?: string[];
}

function rng(seed: string): () => number {
  let h = crypto.createHash("sha256").update(seed).digest();
  let i = 0;
  return () => {
    if (i >= h.length - 4) {
      h = crypto.createHash("sha256").update(h).digest();
      i = 0;
    }
    const v = h.readUInt32BE(i) / 0xffffffff;
    i += 4;
    return v;
  };
}

export function generateFromSchema(schema: Record<string, unknown>, rand: () => number, ctx: { node: string; item?: number; depth?: number; key?: string }): unknown {
  const depth = ctx.depth ?? 0;
  if ("x-mock" in schema) return schema["x-mock"];
  if (Array.isArray(schema.examples) && schema.examples.length) return schema.examples[Math.floor(rand() * schema.examples.length)];
  if (Array.isArray(schema.enum) && schema.enum.length) return schema.enum[Math.floor(rand() * schema.enum.length)];
  if ("const" in schema) return schema.const;
  const type = Array.isArray(schema.type) ? schema.type[0] : schema.type;
  switch (type) {
    case "object": {
      const out: Record<string, unknown> = {};
      const props = (schema.properties ?? {}) as Record<string, Record<string, unknown>>;
      for (const [k, sub] of Object.entries(props)) out[k] = generateFromSchema(sub, rand, { ...ctx, depth: depth + 1, key: k });
      return out;
    }
    case "array": {
      const itemSchema = (schema.items ?? { type: "string" }) as Record<string, unknown>;
      const min = typeof schema.minItems === "number" ? schema.minItems : 1;
      const max = typeof schema.maxItems === "number" ? schema.maxItems : Math.max(min, 3);
      const n = depth > 3 ? min : min + Math.floor(rand() * (max - min + 1));
      return Array.from({ length: n }, (_, i) => generateFromSchema(itemSchema, rand, { ...ctx, depth: depth + 1, key: `${ctx.key ?? "item"}[${i}]` }));
    }
    case "number":
    case "integer": {
      const min = typeof schema.minimum === "number" ? schema.minimum : 0;
      const max = typeof schema.maximum === "number" ? schema.maximum : type === "integer" ? 10 : 1;
      const v = min + rand() * (max - min);
      return type === "integer" ? Math.round(v) : Number(v.toFixed(3));
    }
    case "boolean":
      return rand() > 0.3;
    case "null":
      return null;
    default: {
      const key = ctx.key ?? "value";
      const fmt = schema.format;
      if (fmt === "uri" || /url|link|source/i.test(key)) return `https://example.org/${ctx.node}/${key.replace(/[^a-z0-9]/gi, "")}-${Math.floor(rand() * 1000)}`;
      if (fmt === "date-time") return new Date(1_700_000_000_000 + Math.floor(rand() * 1e10)).toISOString();
      const words = ["evidence", "signal", "claim", "market", "latency", "graph", "reducer", "verifier", "budget", "topology"];
      const pick = () => words[Math.floor(rand() * words.length)];
      return `mock ${key} from ${ctx.node}${ctx.item !== undefined ? `#${ctx.item}` : ""}: ${pick()} ${pick()} ${pick()}`;
    }
  }
}

export class MockBridge implements Bridge {
  readonly name = "mock";
  constructor(private opts: MockOptions = {}) {}

  describe() {
    return "deterministic schema-driven mock (no tokens spent)";
  }

  async available() {
    return { ok: true };
  }

  async execute(req: AgentRequest, ctx: BridgeContext): Promise<AgentResponse> {
    const start = Date.now();
    const seed = `${this.opts.seed ?? "gren"}:${req.run_id}:${req.node_id}:${req.item_index ?? -1}:${req.attempt}`;
    const rand = rng(seed);
    const latency = (this.opts.latencyMs ?? 40) + Math.floor(rand() * (this.opts.jitterMs ?? 60));
    await new Promise<void>((resolve, reject) => {
      const t = setTimeout(resolve, latency);
      ctx.signal.addEventListener("abort", () => {
        clearTimeout(t);
        reject(new BridgeError("cancelled", "cancelled", false));
      }, { once: true });
    });
    if (this.opts.alwaysFail?.includes(req.node_id)) throw new BridgeError(`mock: node ${req.node_id} configured to always fail`, "transport");
    if (rand() < (this.opts.failRate ?? 0)) throw new BridgeError("mock: injected transient failure", "transport");

    let output: unknown;
    if (req.role === "verify") {
      const kill = rand() < (this.opts.killRate ?? 0.25);
      output = { verdict: kill ? "kill" : "pass", reasons: kill ? ["mock verifier: could not confirm the source supports the claim"] : [], confidence: kill ? 0.6 + rand() * 0.4 : 0.2 + rand() * 0.3 };
    } else if (rand() < (this.opts.invalidRate ?? 0) && !req.repair_errors) {
      output = { _invalid: true, note: "mock: injected schema-invalid output" };
    } else {
      output = generateFromSchema(req.output_schema, rand, { node: req.node_id, item: req.item_index });
    }
    const usage = { input_tokens: 400 + Math.floor(req.prompt.length / 4), output_tokens: 250 };
    return {
      output,
      raw: JSON.stringify(output),
      usage,
      cost_usd: estimateCostUsd(req.model, usage),
      model: req.model,
      bridge: this.name,
      duration_ms: Date.now() - start,
      source: "mock",
    };
  }
}

File: gren/gren/models/mock.py
"""Mock provider: deterministic, schema-driven outputs with latency and failure injection. No tokens.

Schema hints: "examples": [..] pick one (seeded); "x-mock": value; "enum": pick one.
Verify calls (schema with a verdict enum) kill with probability `kill_rate`.
"""
from __future__ import annotations

import asyncio
import hashlib
import random
from typing import Any

from ._base import OneShotModel
from .pricing import estimate_cost_usd

WORDS = ["evidence", "signal", "claim", "market", "latency", "graph", "reducer", "verifier", "budget", "topology"]


def generate_from_schema(schema: dict[str, Any], rng: random.Random, node: str, key: str = "value", depth: int = 0) -> Any:
    if "x-mock" in schema:
        return schema["x-mock"]
    if schema.get("examples"):
        return rng.choice(schema["examples"])
    if schema.get("enum"):
        return rng.choice(schema["enum"])
    if "const" in schema:
        return schema["const"]
    if "anyOf" in schema:
        options = [o for o in schema["anyOf"] if o.get("type") != "null"] or schema["anyOf"]
        return generate_from_schema(options[0], rng, node, key, depth)
    t = schema.get("type")
    t = t[0] if isinstance(t, list) else t
    if t == "object":
        return {k: generate_from_schema(sub, rng, node, k, depth + 1) for k, sub in (schema.get("properties") or {}).items()}
    if t == "array":
        items = schema.get("items") or {"type": "string"}
        lo = int(schema.get("minItems", 1))
        hi = int(schema.get("maxItems", max(lo, 3)))
        n = lo if depth > 3 else rng.randint(lo, hi)
        return [generate_from_schema(items, rng, node, f"{key}[{i}]", depth + 1) for i in range(n)]
    if t in ("number", "integer"):
        lo = float(schema.get("minimum", 0))
        hi = float(schema.get("maximum", 10 if t == "integer" else 1))
        v = lo + rng.random() * (hi - lo)
        return round(v) if t == "integer" else round(v, 3)
    if t == "boolean":
        return rng.random() > 0.3
    if t == "null":
        return None
    fmt = schema.get("format")
    if fmt == "uri" or any(w in key.lower() for w in ("url", "link", "source")):
        return f"https://example.org/{node}/{key.replace('[', '-').replace(']', '')}-{rng.randint(0, 999)}"
    if fmt == "date-time":
        return "2026-01-01T00:00:00Z"
    s = f"mock {key} from {node}: {rng.choice(WORDS)} {rng.choice(WORDS)} {rng.choice(WORDS)}"
    while len(s) < int(schema.get("minLength", 0)):
        s += f" {rng.choice(WORDS)} {rng.choice(WORDS)}"
    if "maxLength" in schema:
        s = s[: int(schema["maxLength"])]
    if schema.get("pattern") == "^[a-z0-9-]+$":
        s = "-".join(w for w in s.lower().replace(":", "").split() if w.isalnum())
    return s


class MockModel(OneShotModel):
    def __init__(self, model_id: str = "claude-haiku-4-5", latency_ms: int = 20, jitter_ms: int = 30, fail_rate: float = 0.0,
                 invalid_rate: float = 0.0, kill_rate: float = 0.25, seed: str = "gren", always_fail: list[str] | None = None, **_: Any):
        super().__init__(model_id, latency_ms=latency_ms, jitter_ms=jitter_ms, fail_rate=fail_rate, invalid_rate=invalid_rate,
                         kill_rate=kill_rate, seed=seed, always_fail=always_fail or [])
        self.node_id = "node"
        self.attempt = 1
        self.item: int | None = None
        self.repair = False

    def bind(self, node_id: str, attempt: int, item: int | None, repair: bool) -> None:
        self.node_id, self.attempt, self.item, self.repair = node_id, attempt, item, repair

    async def complete(self, prompt: str, system_prompt: str | None, output_schema: dict[str, Any] | None) -> dict[str, Any]:
        c = self.config
        seed = f"{c['seed']}:{self.node_id}:{self.item}:{self.attempt}:{hashlib.sha1(prompt.encode()).hexdigest()[:8]}"
        rng = random.Random(seed)
        await asyncio.sleep((c["latency_ms"] + rng.random() * c["jitter_ms"]) / 1000)
        if self.node_id in c["always_fail"]:
            raise RuntimeError(f"mock: node {self.node_id} configured to always fail")
        if rng.random() < c["fail_rate"]:
            raise RuntimeError("mock: injected transient failure")
        usage = {"inputTokens": 400 + len(prompt) // 4, "outputTokens": 250}
        cost = estimate_cost_usd(c["model_id"], usage)
        if output_schema is None:
            return {"text": f"mock reply from {self.node_id}", "json": None, "usage": usage, "cost_usd": cost}
        props = output_schema.get("properties") or {}
        is_verify = "verdict" in props and "reasons" in props
        if is_verify:
            kill = rng.random() < c["kill_rate"]
            data: Any = {"verdict": "kill" if kill else "pass",
                         "reasons": ["mock verifier: could not confirm the source supports the claim"] if kill else [],
                         "confidence": round(0.6 + rng.random() * 0.4 if kill else 0.2 + rng.random() * 0.3, 3)}
        elif rng.random() < c["invalid_rate"] and not self.repair:
            data = {"_invalid": True, "note": "mock: injected schema-invalid output"}
        else:
            data = generate_from_schema(output_schema, rng, self.node_id)
        return {"text": "", "json": data, "usage": usage, "cost_usd": cost}
